target_postgres/connector.py

```diff
@@ -277,6 +277,16 @@ def pick_individual_type(jsonschema_type: dict):
         if "object" in jsonschema_type["type"]:
             return JSONB()
         if "array" in jsonschema_type["type"]:
+            # FIXME: This currently uses a non-conformant definition for the Singer SCHEMA.
+            # {"type": "array", "items": {"type": "number"}, "storage": {"type": "vector", "dim": 4}}
+            if (
+                "storage" in jsonschema_type
+                and "type" in jsonschema_type["storage"]
+                and jsonschema_type["storage"]["type"] == "vector"
+            ):
+                from pgvector.sqlalchemy import Vector
+
+                return Vector(jsonschema_type["storage"]["dim"])
```

Comment:
## Thoughts

Within the proposal at https://github.com/meltano/sdk/issues/2102, @edgarrmondragon mentioned an `sqlalchemy_type` attribute within a "catalog metadata in Meltano syntax".

As I am not completely firm with all the terms yet, I don't know if this patch is reasonable to implement the "bottom half" of this story, modulo adding a _dedicated_ type to the typesystem.

This implementation variant deliberately introduces a new `storage` attribute [^1] (instead of `sqlalchemy_type`) within the Singer SCHEMA describing a corresponding dataset, because, well, the target storage system may not be SQL/SQLAlchemy-associated at all, right?

I am sure this implementation variant will never be merged, because, well, reality is probably more complex, and I am just a beginner here. Specifically harmonizing the type system with SQLAlchemy matters vs. reality is something I can't oversee.

## References

- https://github.com/meltano/sdk/issues/2102
- https://github.com/crate-workbench/meltano-target-cratedb/issues/5


[^1]: The flaw probably is here, that "one just doesn't introduce a new attribute" to a Singer SCHEMA, as it will need to run through all sorts of rfc or standardization processes?

Reply:
After getting rid of the very custom `storage` attribute, which naturally did not get respected by any of the Singer SDK code, I figured out that `additionalProperties` might work better instead. See https://github.com/MeltanoLabs/target-postgres/pull/251#pullrequestreview-1780657446.